src/main/anovos/data_ingest/data_ingest.py

```diff
@@ -41,13 +55,129 @@ def read_dataset(spark, file_path, file_type, file_configs={}):
+        This argument determines the cardinality ratio for every column. The cardinality ratio is defined
+        by number of non-null unique values divided by number of non-null total values in a column.
+        If the cardinality ratio is larger than threshold_ratio, its DataType will be treated as Double.
+        Otherwise its DataType will be treated as String. (Default value = 0.005)
+    threshold_string
+        This argument determines the string ratio for every column. The string ratio is defined
+        by number of non-null string values divided by number of non-null total values in a column.
+        If the string ratio is smaller than threshold_ratio, its DataType will be treated as Double.
+        Otherwise its DataType will be treated as String. (Default value = 0.5)
 
     Returns
     -------
     DataFrame
 
     """
     odf = spark.read.format(file_type).options(**file_configs).load(file_path)
+    if parquet_conversion:
+        if not parquet_output_path:
+            raise TypeError(
+                "parquet_output_path cannot be blank for parquet conversion"
+            )
+    if treatment:
+        if id_cols:
+            for col in id_cols:
+                if col not in odf.columns:
+                    raise TypeError("Invalid input for id_cols: " + col)
+        if type(threshold_num) != int:
+            raise TypeError("Invalid input for threshold_num")
+        if type(threshold_ratio) != float or threshold_ratio < 0 or threshold_ratio > 1:
+            raise TypeError("Invalid input for threshold_ratio")
+        if (
+            type(threshold_string) != float
+            or threshold_string < 0
+            or threshold_string > 1
+        ):
+            raise TypeError("Invalid input for threshold_string")
+        list_of_cols = list(c for c in odf.columns if c not in id_cols)
+        odf = odf.persist(pyspark.StorageLevel.MEMORY_AND_DISK)
+        odf_recast = odf.select(
+            [F.col(c).cast(T.DoubleType()).alias(c) for c in list_of_cols]
+        )
+        funs_count = [F.count]
+        exprs_count = [f(F.col(c)) for f in funs_count for c in list_of_cols]
+        list_count_recast = (
+            odf_recast.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+        )
+
+        list_count = odf.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+
+        funs_distinct = [F.countDistinct]
+        exprs_distinct = [f(F.col(c)) for f in funs_distinct for c in list_of_cols]
+        list_distinct_count = (
+            odf.select(*exprs_distinct).rdd.flatMap(lambda x: x).collect()
+        )
+        list_ratio = [i / j for i, j in zip(list_distinct_count, list_count)]
+        list_cat_ratio = [i / j for i, j in zip(list_count_recast, list_count)]
+
+        list_schema = []
+        if id_cols:
+            for col in id_cols:
+                list_schema.append(T.StructField(str(col), T.StringType(), True))
+        for k in range(0, len(list_count)):
```

Comment:
What exactly is going on here? It looks like you're guessing the appropriate type based on the number of distinct values in a column?

Reply:
To me, this kind of heuristic seems quite dangerous and incredibly difficult to debug. Why don't you have the user specify the schema and/or give them the option to override the type for certain columns?

Reply:
The general idea that we are trying to make here is to introduce a somewhat intelligent function to automatically check   and treat the schema of any DataFrame. We can also do as you suggest, but I think it would defeat the purpose of developing some intelligent logic for the treatment.

Reply:
I'm not convinced. This kind of heuristic is very brittle and will have to be re-tuned for every new dataset, and for many datasets you won't find thresholds that yield the correct type for every single column. Just imagine the issues that arise when you're using Anovos to process a daily batch of new data and then one day the statistics change slightly, silently causing the column types to change.

Instead of trying to guess/determine the "correct" thresholds, it will be much easier and - most importantly - more reliable for users to just write the desired schema.

Reply:
I think we already have a function exactly for that, `recast_column` (https://github.com/anovos/anovos/blob/a94b847408143f430b40bc8b0abc35fa1521dea5/src/main/anovos/data_ingest/data_ingest.py#L322). It allows the user to make change to the schema and cast columns as they want. It would be duplicate to do that here.

Reply:
Then why do you need the functionality proposed in this PR in the first place? :-)